preprocess/b64filter_main.cc

```diff
@@ -0,0 +1,137 @@
+
+		// Decoded document buffer
+		string doc;
+
+		for (StringPiece line : in) {
+			base64_decode(line, doc);
+
+			// Description of the document
+			Document document;
+			document.has_trailing_newline = doc.back() == '\n';
+
+			// Make the the document end with a new line. This to make sure
+			// the next doc we send to the child will be on its own line and the
+			// line_cnt is correct.
+			if (!document.has_trailing_newline)
+				doc.push_back('\n');
+
+			document.line_cnt = count(doc.cbegin(), doc.cend(), '\n');
+			
+			// Send line count first to the reader, so it can start reading as
+			// soon as we start feeding the document to the child.
+			line_cnt_queue.Produce(document);
+
+			// Feed the document to the child.
+			// Might block because it can cause a flush.
+			child_in << doc;
+		}
+
+		// Tell the reader to stop
+		line_cnt_queue.Produce(Document{
+			.line_cnt = 0,
+			.has_trailing_newline = false
+		});
+
+		// Flush (blocks) & close the child's stdin
+		child_in.flush();
+		child_in_fd.reset();
+	});
+
+	thread reader([&child_out_fd, &line_cnt_queue]() {
+		util::FileStream out(STDOUT_FILENO);
+		util::FilePiece child_out(child_out_fd.release());
+
+		size_t doc_cnt = 0;
+		Document document;
+		string doc;
+
+		while (line_cnt_queue.Consume(document).line_cnt > 0) {
+			++doc_cnt;
+
+			doc.clear();
+			doc.reserve(document.line_cnt * 4096); // 4096 is not a typical line length
+
+			try {
+				while (document.line_cnt-- > 0) {
+					StringPiece line(child_out.ReadLine());
+					doc.append(line.data(), line.length());
+
+					// ReadLine eats line endings. Between lines we definitely
+					// need to add them back. Whether we add the last one depends
+					// on whether the original document had a trailing newline.
+					if (document.line_cnt > 0 || document.has_trailing_newline)
+						doc.push_back('\n');
+				}
+			} catch (util::EndOfFileException &e) {
+				UTIL_THROW(util::Exception, "Sub-process stopped producing while expecting more lines while processing document " << doc_cnt);
+			}
+
+			string encoded_doc;
+			base64_encode(doc, encoded_doc);
+			out << encoded_doc << '\n';
+
+			// Just to check, next time we call Consume(), will we block? If so,
+			// that means we've caught up with the producer. However, the order
+			// the producer fills line_cnt_queue is first giving us a new line-
+			// count and then sending the input to the sub-process. So if we do
+			// not have a new line count yet, the sub-process definitely can't
+			// have new output yet, and peek should block and once it unblocks
+			// we expect to have that line-count waiting. If we still don't,
+			// then what is this output that is being produced by the sub-
+			// process?
+			if (line_cnt_queue.Empty()) {
+				// If peek throws EOF now our sub-process stopped before its
+				// stdin was closed (producer produces the poison before it
+				// closes the sub-process's stdin.)
+				child_out.peek();
```

Comment:
What about
1) line_cnt_queue.Empty() because we've consumed everything before poison
2) producer produces poison
3) producer closes process stdin
4) peek() throws EOF
5) I think the process dies because nothing catches it.

Reply:
I think you're right. That would be a feasible condition in which the child consumer (whut?) should continue. Any suggestions on how to deal with it? Maybe by changing Empty() into a non-consuming Peek() and checking whether it isn't the poison? Can I do that without a lock?

This whole mechanism assumes that the consumer is quicker than the producer anyway, which isn't often. (However it will look that way if it is producing more output than expected, since that output will be there waiting for us in a buffer…)

Reply:
I guess one could do a non-blocking read.  

But overall I'd rather have a concurrent program I understand rather than one that tries to be fancy.

Reply:
I've simplified the check to just be at the end, after the poison has already been consumed. There the peek() will block till child is closed and then either return happy with input (after producer poison, so bad) or fail on EOF (which is expected).

I'd like to keep the underproducing and overproducing checks in b64filter as these have been the source of multiple issues in the paracrawl pipeline already.